README
 Incorporated Research Institutions for Seismology (IRIS)
 Data Management Center (DMC)
 Data Products Team
 horizontal to vertical spectral ratio (HVSR)

 2017-11-16

----------------------------------------------------------------------------------------------------------------------------------------

 DESCRIPTION:

 The horizontal to vertical spectral ratio (HVSR) of ambient noise provides information on the fundamental natural frequency of the local sediments when
 there is a large acoustic impedance contrast between sediments and underlying rocks.

 This IRIS DMC Python script bundle provides tools to compute and plot horizontal to vertical spectral ratio (HVSR) curves using MUSTANG PDF-PSD web service.
 The bundle contains 2 Python scripts: 

   - computeStationChannelBaseline.py is a Python script that uses MUSTANG's noise-pdf (http://service.iris.edu/mustang/) web service to 
     compute channel-specific noise-baseline for a given station (McNamara et al., 2009). The resulting baseline represents the long-term PSD PDF 
     characteristics of the station-channel pair in the form of median, lower and higher percentiles of the available PSDs for the station.

 BUNDLE CONTENTS:

    CHANGES.txt
       - a text file containing the history of changes to this bundle

    INSTALL.txt
       - installation notes

    README
       - this file

    bin/
       - scripts directory containing:
            + computeStationChannelBaseline.py (described above)
   
    param/
       - parameters directory containing:
            + getStationChannelBaseline_param.py - the configuration parameter file for the computeStationChannelBaseline.py script above

    lib/
       - bundle library files:
            + fileLib.py - a collection of functions to work with files and directories
            + msgLib.py - a collection of functions to print messages

 INSTALLATION:

    see the INSTALL.txt file


USAGE:
   
   bin/computeStationChannelBaseline.py
       Call the script and provide the run arguments. By defining a parameter on the command line, user overrides the values defined for that parameter in the parameter file. 
       The call should have the form:

             getStationChannelBaseline.py net=netName sta=staName loc=locCode {chan=chanCode} start=2007-03-19 end=2008-10-28 {plot=[0,1] verbose=[0,1] percentlow=[10] percenthigh=[90] xtype=[period,frequency]}
         
             Note:
       
             the default values for the parameters between {} may be provided in the parameter file

EXAMPLES:


             getStationChannelBaseline.py net=IU sta=ANMO loc=00 chan=BHZ,BH1,BH2 start=2002-11-19 end=2008-11-13 plot=1 verbose=0 percentlow=10 percenthigh=90
             getStationChannelBaseline.py net=IU sta=ANMO loc=00 start=2002-11-19 end=2008-11-13 plot=1
             getStationChannelBaseline.py net=IU sta=ANMO loc=00 chan=BH1 start=2002-11-19 end=2008-11-13 plot=1


REFERENCES:

McNamara D.E., C.R. Hutt, L.S. Gee, H.M. Benz, and R.P. Buland, 2009, A method to establish seismic noise baselines for automated station assessment, Seismological Research Letters Jul 2009, 80 (4) 628-637; 
DOI: 10.1785/gssrl.80.4.628.  http://srl.geoscienceworld.org/content/gssrl/80/4/628.full.pdf

 HISTORY
    - 2017-11-16: initial release R.2017320
 
 COMMENTS/QUESTIONS:

    Please contact manoch@iris.washington.edu


CHANGES.txt
R.1.3 Release

bin/computeHVSR.py
    2020-06-04 IRIS DMC Product Team (Manoch): V.2020.156, addressed the check mark character display issue on Windows.

bin/getStationChannelBaseline.py
    2020-06-04 IRIS DMC Product Team (Manoch): V.2020.156, addressed the warning about Adding an axes using the same

========================================                                                arguments as a previous axes currently reuses the earlier instance.
R.1.3 Release

bin/computeHVSR.py
    2020-08-27 IRIS DMC Product Team (Manoch): V.2020.240, fixed channel sorting order for horizontal 1 & 2 directions.

INSTALL.txt
    2020-06-04 minor edits

========================================
R.1.2 Release

bin/computeHVSR.py
    2020-06-03 IRIS DMC Product Team (Manoch): V.2020.155, addressed the UTF-8 character issue on Windows.

lib/msgLib.py
    2020-06-03 IRIS DMC Product Team (Manoch): style update V.2020.155

INSTALL.txt
    2020-06-03 Upgrade to Python 3.7

========================================

R.1.1 Release

param/computeHVSR_param.py
    2020-02-24 IRIS DMC Product Team (Manoch): added reportDirectory V.2020.055.

bin/computeHVSR.py
    2020-02-24 IRIS DMC Product Team (Manoch): V.2020.055, now the peak report is also writen to a file under the
                                               data/report directory.

lib/fileLib.py
    2020-02-24  IRIS DMC Product Team (Manoch): style update V.2020.055
========================================

R.1.0 Release

bin/getStationChannelBaseline.py
     2019-06-19 IRIS DMC Product Team (Manoch): V.2019.171, added Peterson 1993 NLNM and NHNM to the plots and updated
                                                color scale.
     2019-06-03 IRIS DMC Product Team (Manoch): V.2019.154, addressed the issue with start and end dates to be the same,
                                                updated the code style, improved plotting.    
bin/computeHVSR.py
    2019-06-19 IRIS DMC Product Team (Manoch): V.2019.171, added Peterson 1993 NLNM and NHNM to the plots and updated
                                               color scale.
    2019-06-03 IRIS DMC Product Team (Manoch): V.2019.154, updated the code style

param/getStationChannelBaseline_param.py
    2019-07-31 IRIS DMC Product Team (Manoch): style update V.2019.212

param/computeHVSR_param.py
    2019-07-31 IRIS DMC Product Team (Manoch): style update V.2019.212

========================================
bin/getStationChannelBaseline.py
    2018-07-10 IRIS DMC Product Team (Manoch): prerelease R.2018191
    2017-11-16 IRIS DMC Product Team (Manoch): R.2017320
    2017-03-12 IRIS DMC Product Team (Manoch): R.2017071

bin/computeHVSR.py
    2018-07-10 IRIS DMC Product Team (Manoch): prerelease R.2018191
    2017-11-28 IRIS DMC Product Team (Manoch): R.2017332
    2017-05-21 IRIS DMC Product Team (Manoch): R.2017141
    2017-03-12 IRIS DMC Product Team (Manoch): R.2017071

lib/fileLib.py
    2018-07-10 IRIS DMC Product Team (Manoch): prerelease R.2018191
    2015-05-19 IRIS DMC Product Team (Manoch): R.2015139

lib/msgLib.py
    2018-07-10 IRIS DMC Product Team (Manoch): prerelease R.2018191
    2015-05-19 IRIS DMC Product Team (Manoch): R.2015139

param/getStationChannelBaseline_param.py
    2018-07-10 IRIS DMC Product Team (Manoch): prerelease R.2018191
    2017-11-16 IRIS DMC Product Team (Manoch): R.2017320
    2017-03-12 IRIS DMC Product Team (Manoch): R.2017071

param/computeHVSR_param.py
    2018-07-10 IRIS DMC Product Team (Manoch): prerelease R.2018191
    2018-06-18 IRIS DMC Product Team (Manoch): added removeOutliers parameter to allow HVSR computation without 
                                               removing outliers and added method parameter that indicates the method to use for combining h1 and h2
    2017-11-28 IRIS DMC Product Team (Manoch): R.2017332
    2015-05-19 IRIS DMC Product Team (Manoch): R.2017139


lib/msgLib.py
#
# DESCRIPTION
#   a collection of functions to print messages
#
# HISTORY
#   2020-06-03 IRIS DMC Product Team (Manoch): style update V.2020.155
#   2018-07-10 IRIS DMC Product Team (Manoch): pre-release V.2018.191
#   2015-05-19 IRIS DMC Product Team (Manoch): created (V.2015.139)
#
# NOTES
# 
#


def message(post_message):
    """print a run message"""
    bar = "*" * 12
    print("%s %s %s" % (bar, post_message, bar))


def error(err_message, code):
    """print an error message"""
    print("\n[ERR] %s\n" % err_message, flush=True)
    return code


def warning(sender, warn_message):
    """print a warning message"""
    print("[WARN] from %s: %s" % (sender, warn_message), flush=True)


def info(info_message):
    """print an informative message"""
    print("[INFO] %s" % info_message, flush=True)
    return


def param(params, var):
    """get a variable from the parameter file by name"""
    if var in dir(params):
        return params
    code = error("'".join(["variable ", var, " not in the parameter file"]), code=2)
    import sys
    sys.exit(code)


lib/fileLib.py
"""
  DESCRIPTION
    a collection of functions to work with files and directories

  HISTORY
    2020-02-24 IRIS DMC Product Team (Manoch): style update V.2020.055
    2019-06-03 IRIS DMC Product Team (Manoch): release V.2019.154
    2018-07-10 IRIS DMC Product Team (Manoch): pre-release V.2018.191
    2015-05-19 IRIS DMC Product Team (Manoch): created (V.2015.139)
"""

import os
import sys

hvsr_directory = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
library_path = os.path.join(hvsr_directory, 'lib')
sys.path.append(library_path)

import msgLib


def mkdir(base_directory, target_directory):
    """ Make a directory if it does not exist."""
    try:
        _directory = os.path.join(base_directory, target_directory)
        if not os.path.exists(_directory):
            os.makedirs(_directory)
        return _directory
    except Exception as _er:
        return msgLib.error(f'failed to create directory {_directory}\n{_er}', None)


def baselineFileName(network, station, location, channel):
   """Construct the baseline file name."""
   label = '.'.join([network, station, location, channel, 'txt'])
   return label


def hvsrFileName(network, station, location, start, end):
   """Construct the HVSR file name."""
   label = '.'.join([network, station, location, start, end, 'HVSR', 'txt'])
   return label


param/getStationChannelBaseline_param.py
"""
IRIS HVSR bundle

#DESCRIPTION:
  HVSR configuration parameters for getStationChannelBaseline

  HISTORY
    2019-07-31 IRIS DMC Product Team (Manoch): style update V.2019.212
    2019-06-03 IRIS DMC Product Team (Manoch): Release V.2019.154
    2018-07-10 IRIS DMC Product Team (Manoch): pre-release V.2018-191
    2017-11-16 IRIS DMC Product Team (Manoch): (V.2017.320)
    2017-03-12 IRIS DMC Product Team (Manoch): created (V.2017.071)
"""

import os
import sys

# The parent directory is one level up.
parentDirectory = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
libraryPath = os.path.join(parentDirectory, 'lib')
sys.path.append(libraryPath)

import fileLib as fileLib

# Print more information during execution [0=no, 1=yes].
verbose = 0
print('[INFO] libraryPath: {}'.format(libraryPath))


# MUSTANG service URL.
mustangUrl = 'http://service.iris.edu/mustang/noise-pdf/1/query?'

# Default directory paths.
dataDirectory = fileLib.mkdir(parentDirectory, 'data')
imageDirectory = fileLib.mkdir(parentDirectory, os.path.join('image', 'baseline'))
workDir = fileLib.mkdir(parentDirectory, 'scratch')
baselineDirectory = fileLib.mkdir(dataDirectory, 'baseline')

# Default station channel codes.
chan = 'BHZ,BH1,BH2'

# X-axis  type ['period', 'frequency'].
xtype = 'period'

# Percentiles to compute.
percenthigh = 90
percentmid = 50
percentlow = 5


# Plot parameters:

# Plot the values [0=no, 1=yes]
plot = 0
# Plot the New High/Low Noise Model see [Peterson1993] [0=no, 1=yes]
plotnnm = 0

# Type of plot for the percent values ['line', 'scatter']
percentPlotType = 'line'


# Axes label.
yLabel = 'Power (dB)'
xLabel = {'frequency':'Frequency (Hz)', 'period':'Period (s)'}

# Default x-axis limits for the two possible xtype above ['period', 'frequency'].
xLim = {'frequency': [0.01, 9], 'period': [0.15, 200]}

# Default Y-axis limits.
yLim = [-200, -50]

# Probability limits.
pMin = -0.3
pMax = 30

# Color of the percentile markers/lines.
colorLow = 'black'
colorMid = 'yellow'
colorHigh = 'red'

# Plot resolution.
imageDpi = 150

# Plot size.
imageSize = [4.4, 3.3]

# Value plot transparency.
alpha = 0.3

# Line width for the lines/markers.
lw = 2

# Line width for percentile values and NLM.
lw2 = 1


param/computeHVSR_param.py
"""
 IRIS HVSR

 DESCRIPTION
 computeHVSR.py configuration parameters

 HISTORY
  2020-02-24 IRIS DMC Product Team (Manoch): added reportDirectory V.2020.055
  2019-07-31 IRIS DMC Product Team (Manoch): style update V.2019.212
  2018-07-10 IRIS DMC Product Team (Manoch): pre-release V.2018.191
  2018-06-18 IRIS DMC Product Team (Manoch): added removeOutliers parameter to allow HVSR computation without 
                      removing outliers and added method parameter that indicates the method to use for 
                      combining h1 and h2
  2017-11-28 IRIS DMC Product Team (Manoch): V.2017332
  2015-05-19 IRIS DMC Product Team (Manoch): created R.2017139

 NOTES

"""
import os
import sys

parentDirectory = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))

# import the libraries
libraryPath = os.path.join(parentDirectory, 'lib')
sys.path.append(libraryPath)

import fileLib as fileLib

verbose = 1

# URLs
mustangPsdUrl = 'http://service.iris.edu/mustang/noise-psd/1/query?'
mustangPdfUrl = 'http://service.iris.edu/mustang/noise-pdf/1/query?'

# Directories
dataDirectory = fileLib.mkdir(parentDirectory, 'data')
imageDirectory = fileLib.mkdir(parentDirectory, os.path.join('image', 'hvsr'))
workDir = fileLib.mkdir(parentDirectory, 'scratch')
baselineDirectory = fileLib.mkdir(dataDirectory, 'baseline')
reportDirectory = fileLib.mkdir(dataDirectory, 'report')
hvsrDirectory = fileLib.mkdir(dataDirectory, 'hvsr')

# Default station channel list.
chan = 'BHZ,BHN,BHE'

# Should break the long requested interval to 'n' segments for smaller data request chunks.
# Increase 'n' when requesting long time intervals to avoid request failures.
n = 1

# Define x-axis  type (period or frequency).
xtype = 'frequency'
hvsrband = [0.2, 15]

# should we remove outliers? (0|1)
removeoutliers = 1

# minimum peak amplitude to be considered. By increasing the waterlevel, you will reduce the sensitivity of
# peak picking.
waterlevel = 1.8

# minimum rank to be accepted
minrank = 2

# H is computed based on the selected method for combining h1 and h2
#     see: https://academic.oup.com/gji/article/194/2/936/597415 FOR METHODS 2-6
#     method:
#        (1) Diffuse Field Assumption
#        (2) arithmetic mean H = (HN + HE)/2 (the horizontal components are combined by using a simple mean)
#        (3) geometric mean H = Sqrt(HN . HE) (mean horizontal spectra is derived by taking square-root of the
#            product of the two horizontal components)
#        (4) vector summation H = Sqrt(N^2 + E^2)
#        (5) quadratic mean H = Sqrt((N^2 + E^2)/2.0)
#        (6) maximum horizontal value H = Max(HN, HE)
methodList = ['', 'Diffuse Field Assumption', 'arithmetic mean', 'geometric mean', 'vector summation',
              'quadratic mean', 'maximum horizontal value']

method = 4

# plot
plot = 1
plotnnm = 1
showplot = 1
plotpsd = 0
plotpdf = 1
plotbad = 0
yLabel = 'Power (dB)'
xLabel = {'frequency': 'Frequency (Hz)', 'period': 'Period (s)'}
xLim = {'frequency': [0.001, 20], 'period': [0.1, 120]}
yLim = [-200,-50]
#yLim = [-150,-70]
#yLim = [-200,-100]

hvsrylim = [0,5]
hvsrXlim = hvsrband
hvsrYlabel = 'HVSR'


pMin = -0.5
pMax = 30

colorLow = 'blue'
colorMid = 'yellow'
colorHigh = 'red'
imageDpi = 150
imageSize = [8, 16]
alpha = 0.7
lw = 3
